testing/src/test/scala/org/http4s/testing/ErrorReporting.scala

```diff
@@ -9,11 +9,81 @@
 
 package org.http4s.testing
 import java.io.{ByteArrayOutputStream, PrintStream}
-
+import cats.implicits._
+import cats.Monad
+import org.http4s.{Headers, MessageFailure, Request, Response, Status}
+import org.http4s.headers.{Connection, `Content-Length`}
+import org.http4s.syntax.string._
 import scala.util.control.NonFatal
+import java.net.URL
+import java.security.{
+  AccessControlContext,
+  CodeSource,
+  Permissions,
+  PrivilegedAction,
+  ProtectionDomain
+}
 
 object ErrorReporting {
 
+  /**
+    * Silences System.out and System.err streams for the duration of thunk.
+    * Restores back the original streams before exiting.
+    */
+  def silenceOutputStreams[R](thunk: => R): R =
+    synchronized {
+      val originalOut = System.out
+      val originalErr = System.err
+
+      // Create blank permissions
+      val perm = new Permissions
+
+      // CodeSource domain for which these permissions apply (all files)
+      val certs: Array[java.security.cert.Certificate] = null
+      val code: CodeSource = new CodeSource(new URL("file:/*"), certs)
+
+      val domain = new ProtectionDomain(code, perm)
+      val domains = new Array[ProtectionDomain](1)
+      domains(0) = domain
+      val context = new AccessControlContext(domains)
+
+      // Redirect output to dummy stream
+      val fakeOutStream = new PrintStream(new ByteArrayOutputStream)
+      val fakeErrStream = new PrintStream(new ByteArrayOutputStream)
+      System.setOut(fakeOutStream)
+      System.setErr(fakeErrStream)
+      val result = java.security.AccessController.doPrivileged(
```

Comment:
Can this throw?  I'd expect to see a finally block on the reset, but I'm not so familiar on this API.

Reply:
This can throw a `AccessControlException` in the `doPrivileged` block if the code being executed inside it tries to redirect `System.out` or `System.err` output streams. We cannot catch it because then we would have to return some result of type `R`. If there is a code that is trying to fiddle with `System.out` (redirecting out put streams), then `silenceOutputStreams` method cannot be used to surround it with.

Reply:
But even if it throws, the test will error, and we can keep running other tests, right?  I don't think there should be a catch here, but a finally still makes sense to me.  Or are you saying anything we throw from here is to be treated as fatal and invalidates any other tests?

Reply:
What should go in the finally block?

Reply:
Lines 62 and 63.  That way, even if `doPrivileged` throws, the outputs will be restored. the exception will be thrown, the test framework will record it, and we can try to run more tests in a clean state.

Reply:
ah okay, got it. Done